internal/controllers/reference/reference.go

```diff
@@ -0,0 +1,125 @@
+package reference
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane"
+)
+
+const (
+	VersionV1  = "v1"
+	KindSecret = "Secret"
+)
+
+// UpdateReferencesToSecret updates the reference records between referrer and each secret
+// in namespacedNames in record cache.
+func UpdateReferencesToSecret(
+	ctx context.Context,
+	c client.Client, indexers CacheIndexers, dataplaneClient *dataplane.KongClient,
+	referrer client.Object, namespacedNames []types.NamespacedName,
+) error {
+	for _, nsName := range namespacedNames {
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: nsName.Namespace,
+				Name:      nsName.Name,
+			},
+		}
+
+		// Here we update the reference relationship even when the referred secret does not exist yet
+		// If the referred secret is created, it could be reconciled in secret controller.
+		referrerCopy := referrer.DeepCopyObject().(client.Object)
+		if err := indexers.SetObjectReference(
+			referrerCopy, secret.DeepCopy()); err != nil {
+			return err
+		}
+
+		if err := c.Get(ctx, nsName, secret); err != nil {
+			return err
+		}
+
+		if err := dataplaneClient.UpdateObject(secret); err != nil {
+			return err
+		}
+	}
+
+	return removeOutdatedReferencesToSecret(indexers, dataplaneClient, referrer, namespacedNames)
+}
+
+// removeOutdatedReferenceToSecret removes outdated reference records to secrets in reference indexer.
+// secrets that are really referred by referrer is passed in parameter referredSecretNames.
+// if the secret is not actually referrenced by any object after deleting outdated reference records,
+// the secret will be removed from object cache in dataplaneClient.
```

Comment:
Can we used a different wording for this? I'm not sure I fully understand what it was meant to say by let me try suggesting something:

```suggestion
// secrets that are referred by referrer are passed in referredSecretNames parameter.
// if a secret is not referenced by any other object after deleting outdated reference records,
// the secret will be removed from object cache in dataplaneClient.
```

BTW: isn't 

> removed from object cache in dataplaneClient

removing via `dataplaneClient` actually removing this in Kong GW config? 🤔

Reply:
No, the `dataplaneClient` keeps a `storer` to store all the k8s objects (so I call the `storer` as "object cache" or "object store") that are possibly used in Kong GW config. Currently, all `Secret`s are stored in the `storer`. The PR here is submitted to limit the `storer` to store the secrets that are possible to used in Kong GW config.

Reply:
updated comments here.